### ISSUE TITLE
Always getting "100% of new lines are covered"

### ISSUE THREAD

**Issue by acwl-aiq** [py-cov-action/python-coverage-comment-action#11]
Thanks for writing this interesting GH action script. I was experimenting with it in a sandbox repo. The line and branch coverage appears to be working as intended but the new line coverage is always showing 100%. 

I probably did something wrong in the configuration that I wasn't able to immediately notice. I wonder if you can lend an eye on this?

This is a pull request where I deliberately introduced code that's not covered by tests. https://github.com/acwl-aiq/lambda_lock_poc/pull/7

Additionally, is there a config I can use to make decreases in line/branch coverage to show red instead of blue? Thanks!

**Comment by ewjoachim** [comment]
Hm, strange. I don't have an idea right now, I'll need to give it a look later. No promises, but I'll see what I can do.

**Comment by acwl-aiq** [comment]
That's fine. This is not urgent

**Comment by kieferro** [comment]
I have copied the repo and looked at the logs. It seems that the coverage module, when using 
```
source =
        lib
        lambda_lock_poc
        construct
```
in the `.coveragerc` file only writes relative paths regarding the sources into the XML file and the `diff-cover` module then can't handle this and doesn't find the files. With only new lines without coverage information the module outputs 100% coverage. (It might be useful to write in the PR-comment in such a case that no lines with coverage information were created instead of 100%. I could implement that, if you want @ewjoachim ).

I'm not quite sure if there's a bug in there somewhere or if this is all expected behavior, but one workaround I've found is to take the source information out of the .coveragerc file and instead write into the workflow file which folders to test. So instead of this line:
`coverage run --rcfile=.coveragerc -m pytest`
this one:
`coverage run --rcfile=.coveragerc -m pytest ./lib/ ./lambda_lock_poc/ ./tests/`

I tested this with a copy and it worked for me so far. Would this be a satisfactory solution?

Regarding the red arrow, there is no setting. However, I could implement it so that you can specify which emojis should be displayed when the coverage increases/decreases. (Of course, only if @ewjoachim thinks it makes sense to implement this feature).

**Comment by ewjoachim** [comment]
Starting with the easy answer:

> Regarding the red arrow, there is no setting. However, I could implement it so that you can specify which emojis should be displayed when the coverage increases/decreases. (Of course, only if @ewjoachim thinks it makes sense to implement this feature).

I'd say the way I'd be most comfortable with is to take the idea of overriding the template a bit further (with `COMMENT_TEMPLATE` but it would be nice if the base template could act as a real "base" template, with multiple levels of Jinja blocks. This way, we could easily override the template and just replace a single block.)

And I'd be happy to review a PR doing that. Please include the proper documentation so that people know how to use it. Also, if you do, please open a separate issue/PR :) Thanks ❤️  (< that heart is red 😛 ) 

(oh and also, just so that it's said even if it's useless: the fact that the arrow is blue is a choice by either GitHub or your OS, but it could very well be that on someone's setup, the same arrow is green, yellow or purple. The actual color is not specified in the official Unicode description , as far as I know).

---

Now regarding the real issue... The aim would be that people can easily take the most out of this tool, so we need to make it explicit when it's misused, and/or improve documentation to make sure people manage to do it. (In all honesty... I'm wondering if a better "diff-cover" lib, even one we'd do ourselves, would be easier.)

I'm accepting any PR going in that direction, and if you start something, I'll try to do my best to find some time to help you see it merged.

**Comment by kieferro** [comment]
> you can already replace the template with `COMMENT_TEMPLATE`

I had actually assumed that this feature was still a work in progress. When I went through the code to see how it worked, I noticed that while it is defined as an input in `action.yml`, it is then not used anywhere (it is not even passed as an environment variable to the Docker image).

> it would be nice if the base template could act as a real "base" template, with multiple levels of Jinja blocks. This way, we could easily override the template and just replace a single block.)

Just to make sure I understand this correctly: the idea would be to split the template into multiple sections (for example: title, comparison to previous coverage, new lines) and then allow the user to specify their own template for those individual sections?


> The aim would be that people can easily take the most out of this tool, so we need to make it explicit when it's misused, and/or improve documentation to make sure people manage to do it. (In all honesty... I'm wondering if a better "diff-cover" lib, even one we'd do ourselves, would be easier.)

Yes, the specific problem at this point is that it's not really clear to me where the problem lies with this explicit issue. I've never worked with diff-coverage before and I'm not sure if it works the way it's supposed to in this case and if it could be configured differently somehow. I'm also not sure how easy it is to write a custom module that compares the coverage.

> I'm accepting any PR going in that direction, and if you start something, I'll try to do my best to find some time to help you see it merged.

If I have an idea how something could be improved, I open an issue/PR, thanks :)

**Comment by ewjoachim** [comment]
> it is then not used anywhere

Oops... You're probably right. It was a feature in v1 and I probably forgot to port it in v2. That said:

- [Here](https://github.com/ewjoachim/python-coverage-comment-action/blob/v2/coverage_comment/main.py#L120), read from the config instead of calling `template.read_template_file()`
- https://github.com/ewjoachim/python-coverage-comment-action/blob/v2/coverage_comment/settings.py#L21: Add a settings value reading `COMMENT_TEMPLATE` and a `clean_comment_template()` method that, when empty, calls  `template.read_template_file()`

I believe it should be enough to load a custom template :)

> Just to make sure I understand this correctly: the idea would be to split the template into multiple sections (for example: title, comparison to previous coverage, new lines) and then allow the user to specify their own template for those individual sections?

Yes, I don't know if you're familiar with how [jinja `{% block x %}`s and `{% extend y %}` work](https://jinja.palletsprojects.com/en/3.0.x/templates/#template-inheritance), but this means you can have a single template with multiple blocks defined, and then users can either define a new base template or declare they `extend` our base and just define the `{% block %}` they want to override.

If we do this, let's make sure we're using a [SandboxedEnvironment](https://jinja.palletsprojects.com/en/3.0.x/sandbox/) because I'm not sure we can 100% trust the template in this case.

> I've never worked with diff-coverage before and I'm not sure if it works the way it's supposed to in this case and if it could be configured differently somehow

I'm not 100% sure either. From my understanding, diff-coverage reads coverage data, computes the diff from the base branch, and counts how many lines added in the diff appear in the coverage. So it doesn't compute a diff of the coverage (it won't tell you that you went from x% to y%), but it computes coverage on the added lines in the diff (it will tell you that x% of the lines you added are covered) and more specifically, it looks at all the lines that appear both in the coverage report (either as _covered_ or _not covered_) and the lines added in the diff (this way, it doesn't report on README.md as being "not covered" . So from my understanding, it doesn't need specifying a module or what. As said, I didn't spend a lot of time investigating (and I'm sorry about that)

**Comment by kieferro** [comment]
> * Add a settings value reading `COMMENT_TEMPLATE` and a `clean_comment_template()` method that, when empty, calls  `template.read_template_file()`

Yes, I think the implementation of that should be pretty straightforward. I'll be happy to implement it and then add a PR in the next few days.

> Yes, I don't know if you're familiar with how [jinja `{% block x %}`s and `{% extend y %}` work](https://jinja.palletsprojects.com/en/3.0.x/templates/#template-inheritance),

I've worked with Jinja before, but I'm not extremly familiar with it yet. The concept itself seems relatively clear to me and I will have a closer look at it. Thank you very much for the links!

> I'm not 100% sure either. From my understanding, diff-coverage reads coverage data, [...]

Yes, that's how I understand it too

> So from my understanding, it doesn't need specifying a module or what.

I rather thought that it could maybe be somehow possible to specify additional sources for the XML file, for which the relative paths are then used or something like that. The problem in the above case is that it doesn't find the files because the paths for the files in the XML file are relative to the folders specified as sources in `.coveragerc` and not relative to the folder from which the command was executed. This could also be a bug in either `coverage.py` or `diff-coverage`, but I can't tell because I just don't know how they are supposed to work.

**Comment by kieferro** [comment]
Closing this since the ambiguous wording mentioned has been fixed in #95.
If you still have problems, feel free to reopen @acwl-aiq
